### DISCUSSION TITLE
hierarchical pooling methods for different size graph inputs

### DISCUSSION THREAD

**Discussion by mfarazi1991** [pyg-team/pytorch_geometric#2871] [Q&A]
Hi,

I have this very big issue on graph classification of varying size graphs (primarily meshes with 10k+ nodes) using the Cheb_conv method. In all examples, there is just one graph fixed (they assume fixed size graphs) and then use pooling once since the topology is fixed. However, when the inputs are not of fixed size, I cannot figure out how to do pooling on each instance separately. Mainly I have issues with how to batch the input graphs (they have different laplacian and dimensions).

**Comment by rusty1s** [comment]
Can you give a small example where exactly you have issues? In general, you can simply use the `torch_geometric.data.DataLoader` to generate mini-batches out of our examples, and these can be well inputted into `ChebConv` as well. For pooling, all pooling operators support the `batch` vector argument, which is necessary to take care of correctly pooling graphs in mini-batch mode.

**Reply by mfarazi1991** [reply]
```
path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'MNIST')
transform = T.Cartesian(cat=False)
train_dataset = MNISTSuperpixels(path, True, transform=transform)
test_dataset = MNISTSuperpixels(path, False, transform=transform)
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64)
d = train_dataset


def normalized_cut_2d(edge_index, pos):
    row, col = edge_index
    edge_attr = torch.norm(pos[row] - pos[col], p=2, dim=1)
    return normalized_cut(edge_index, edge_attr, num_nodes=pos.size(0))


class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = SplineConv(d.num_features, 32, dim=2, kernel_size=5)
        self.conv2 = SplineConv(32, 64, dim=2, kernel_size=5)
        self.fc1 = torch.nn.Linear(64, 128)
        self.fc2 = torch.nn.Linear(128, d.num_classes)

    def forward(self, data):
        data.x = F.elu(self.conv1(data.x, data.edge_index, data.edge_attr))
        weight = normalized_cut_2d(data.edge_index, data.pos)
        cluster = graclus(data.edge_index, weight, data.x.size(0))
        data.edge_attr = None
        data = max_pool(cluster, data, transform=transform)

        data.x = F.elu(self.conv2(data.x, data.edge_index, data.edge_attr))
        weight = normalized_cut_2d(data.edge_index, data.pos)
        cluster = graclus(data.edge_index, weight, data.x.size(0))
        x, batch = max_pool_x(cluster, data.x, data.batch)

        x = global_mean_pool(x, batch)
        x = F.elu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        return F.log_softmax(self.fc2(x), dim=1)
```

In this example, all of the images and the created graphs from images have the same dimension (number of nodes), therefore we just use pooling for one graph and we calculate Laplacian (I use Chebyshev_Conv so for each coarsened graph I have to calculate Laplacian too) of the coarsened ones. Then we apply it to all of the samples since they all are the same but have different features. However, in my dataset, the dimensions of input graphs are not the same, therefore, for each graph, I have to do pooling. Using this example I get errors but generally, is that even possible? I am not talking about the global mean pooling at the end.

**Reply by rusty1s** [reply]
This should be possible. All pooling operators should work just fine in mini-batch mode. Your example looks correct to me. What is the error message you encounter?

Note that you do not need to (re-)compute Laplacian in case you use `normalization="sym"` within `ChebConv`.

**Reply by mfarazi1991** [reply]
This is the error I get either with default code (your example on Graclus clustering) with MNIST and my dataset.
     86 
     87     if edge_weight is not None:
**---> 88         assert edge_weight.numel() == edge_index.size(1)**
     89         loop_weight = edge_weight.new_full((N, ), fill_value)
     90         edge_weight = torch.cat([edge_weight, loop_weight], dim=0)



Also, in other codes available for ChebNet paper, there is a line of code similar in all of them:


```
L, perm = coarsen(A, coarsening_levels)  --> A is a fixed graph shared among all, in my case A is varying in size
g_arr = [dgl.from_scipy(csr) for csr in L]
```
Therefore, the batching is based on the fixed Ls and perm but in my case these are varying.

**Reply by rusty1s** [reply]
Are you calling `add_self_loops` anywhere in your code? You might wanna check why the shapes of `edge_index` and `edge_weight` do not align in your case. The MNIST Graclus examples provided by PyG runs fine for me.

**Reply by mfarazi1991** [reply]
I found the error. It seems `data.x = F.elu(self.conv1(data.x, data.edge_index))` is true not adding edge_attr to the argument for the latest PyG. I removed it and it worked. But I am not sure why the last Graclus calling does not change the number of nodes (I used some print-out functions to track the number of nodes) and after the last one which comes with max_pool_x instead of max_pool function, does not change the number of nodes at all. Any thoughts? Btw, huge thanks for your swift and helpful response.

**Reply by rusty1s** [reply]
The number of nodes depend on the number of unique values in `cluster`, i.e., `cluster.unique().numel()`. In case this value aligns with the number of output nodes, this indicates that `graclus` cannot downscale the graph any further.
